### PR TITLE
RNA pipeline with adapter clipping

### PR DESCRIPTION
**Pull request by takutosato**
After this change the RNA pipeline will do adapter clipping (using fastp) prior to alignment. This will help to retain small insert reads (which are fairly commonplace in FFPE) that would otherwise be discarded before running RSEM.

- [x] confirm that dockers for the fastqc and fastp tasks meet the standard for a warp pipeline
- [x] update internal documentation (https://docs.google.com/document/d/1VJMw2vS7gpxxO0G4L1KsYU-Whs7Zk2v_J3p4OBsrbsc/edit?usp=sharing)
- [x] merge GATK TransferReadTags
- [x] determine whether RSEM will or will not be run as part of the pipeline
- [x] update the output spreadsheet
- [x] update flowchart
- [x] Check that we are not assuming hg19 in the new clipping code; transcriptome names?

~~Related but not required tasks~~One last task:
- [x] Pre-processing tool for RSEM: https://github.com/broadinstitute/gatk/pull/7752

### CODE REVIEW

**Comment by takutosato** [comment]
@gbggrant @mmorgantaylor @kcibul Now that we convert the files back to fastq whether the input is a ubam or fastqs, the variables platform_unit etc. are no longer optional. Will that cause problems either upstream or downstream of this wdl?

**Comment by takutosato** [comment]
image

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (wjdingman1) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:34]: For the 'parameter meta' for all these values we can remove the 'only required when using fastq files as input blurb'

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (wjdingman1) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:105]: no longer needed, can just use `FastqToUbam.unmapped_bam`

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (wjdingman1) [inline tasks/broad/RNAWithUMIsTasks.wdl:110]: Is this something we may want to be configurable in the future i.e. as top level wdl input? Its not necessary just something to consider

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (wjdingman1) [inline tasks/broad/RNAWithUMIsTasks.wdl:122]: what is this tool and is it regularly maintained/updated? If this docker image is something that is an official release we can leave it as is, otherwise we will want to move it into a warp managed image

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (wjdingman1) [inline tasks/broad/RNAWithUMIsTasks.wdl:931]: Im guessing the functionality from this JAR will eventually be folded into GATK?

**Review by wjdingman1 — COMMENTED** [review]
Looks good on first pass! Let me know whatever help you need with docker stuff.

Also we have a style guide on how we like to format our runtime blocks -> [here](https://github.com/broadinstitute/warp/blob/develop/website/docs/contribution/contribute_to_warp/wdl_task_runtime_style.md)

If you have time it would be appreciated if you modify your tasks to that style, if not, then I can find some time to do it, thanks!

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (wjdingman1) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:26]: Sorry, trying to reference the bam as input

**Comment by wjdingman1** [comment]
looks like womtool is complaining about inputs for `BroadInternalRNAWithUMIs` 
image

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:105]: I don't think so - you can still provide either bam or fastqs as top level inputs. It's just that later in the pipeline he's making fastqs for the fastqc task and so needs the values like platform, etc to be defined.

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:107]: Not sure this comment is really necessary?

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:135]: minor nit. ".adapter_clipped"??

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (gbggrant) [inline tasks/broad/RNAWithUMIsTasks.wdl:131]: Not sure what is meant by these comments.

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (gbggrant) [inline tasks/broad/RNAWithUMIsTasks.wdl:868]: Might as well go to 2.26.11 (which address the log4j vulnerability.)

**Review by gbggrant — CHANGES_REQUESTED** [review]

**Comment by gbggrant** [comment]
@takutosato Thanks - I don't think that making those parameters non-optional will break anything upstream or downstream. We'll need to update a couple of the tests to now supply those tag values (where the inputs are bams).

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (takutosato) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:78]: Sounds good

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (takutosato) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:26]: I think we still want to support both ubam and fastqs as input

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (takutosato) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:34]: Yes

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:110]: I won't change but we will probably need to move it somewhere.

@gbggrant where should it go? All the other resources files are in `gs://gcp-public-data--broad-references/` but it might not be appropriate there since it's not part of the reference.

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:122]: https://github.com/OpenGene/fastp is the tool and the docker comes form biocontainer...definitely not an official release, it was the best I could find.

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:931]: Yes

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (takutosato) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:107]: Removed

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (takutosato) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:135]: fixed

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:131]: removed

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:868]: done

**Comment by takutosato** [comment]
Thank you for reviewing @wjdingman1 @gbggrant. I addressed your comments and updated the runtime blocks to conform to the style guide.

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (wjdingman1) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl]: ohhhh okay that makes sense

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:904]: This is a docker used by TAG team from GP; should I just ask for a docker file from them?

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl]: Hmm. I tested with 0.20.1 so we should stick with that. It should be safe to use 0.23.1 since all the updates since 0.20.1 are not related to adapter trimming, but just to be safe.

Comment (wjdingman1) [reply tasks/broad/RNAWithUMIsTasks.wdl]: okay sounds good!

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (wjdingman1) [reply tasks/broad/RNAWithUMIsTasks.wdl:904]: if its something they regularly maintain and do security scans on then its okay to leave as is, otherwise we'll take care of bringing it to warp

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:78]: You still should keep this task and have it verify that either the bam is defined OR a pair of fastqs are both defined.
You could do so here, in this WDL, or in the tasks wdl (but if you do it in the top-level, then the task should be deleted from the task wdl).

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:77]: Assume what?

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:113]: Bug!
```suggestion
      fastq2 = SamToFastq.fastq2,
```

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [inline pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:155]: clean up this comment please.

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (gbggrant) [reply tasks/broad/RNAWithUMIsTasks.wdl:112]: It's a good question - still will probably put it in the Google public references bucket.

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (gbggrant) [inline tasks/broad/RNAWithUMIsTasks.wdl:939]: Please use "GiB" or "GB" consistently.

**Comment by gbggrant** [comment]
HI @takutosato a couple more comments / minor fixes. Found one bug.
I'll put that adapters file somewhere publicly accessible.

**Comment by wjdingman1** [comment]
@takutosato I added the fastp docker image changes to this PR. 

Unrelated but you need to update the changelog.md to be consistent with the current pipelnie version `1.0.4` and not the changes that you have made in there

File: pipelines/broad/internal/rna_seq/BroadInternalRNAWithUMIs.changelog.md
Comment (wjdingman1) [inline pipelines/broad/internal/rna_seq/BroadInternalRNAWithUMIs.changelog.md:2]: has the wdl pipeline version been updated?

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (wjdingman1) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl:78]: @takutosato just making sure you saw this

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (wjdingman1) [reply tasks/broad/RNAWithUMIsTasks.wdl:912]: @takutosato pinging on this

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl]: Actually, I just added it back in.

File: pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl
Comment (gbggrant) [reply pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl]: @takutosato I fixed this - I'm 99.999999% certain it was a bug, but please verify

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [inline tasks/broad/RNAWithUMIsTasks.wdl:973]: @gbggrant I removed `_gatk` as I thought it wasn't descriptive; did it cause an error without it? Perhaps we should use `_RSEM_post_processed` instead?

Comment (gbggrant) [reply tasks/broad/RNAWithUMIsTasks.wdl:973]: @takutosato I added it back because that's what you are naming the file on line 969 - I guess you should remove it there in addition.

**Review by gbggrant — APPROVED** [review]

**Review by wjdingman1 — APPROVED** [review]

**Comment by wjdingman1** [comment]
@gbggrant Do we want to update plumbing based off what we have here or does it not matter because of the non-deteminism

**Comment by gbggrant** [comment]
@wjdingman1 we'll definitely need to update plumbing truth for this PR. There's some new plumbing data, plus some changes in the pipeline. Looks like this is ready to merge, can update the truth once merged.

**Comment by gbggrant** [comment]
(just waiting for this last smart test to run to completion).

File: tasks/broad/RNAWithUMIsTasks.wdl
Comment (takutosato) [reply tasks/broad/RNAWithUMIsTasks.wdl:897]: @wjdingman1 Sorry for the delay. It's updated about once every year and I doubt they do any security scans. I can share the docker file. It looks like the docker installs extraneous tools, it might be good to create a new docker (not necessarily in this PR)---could I get your help trimming it down?

**Pull request merged**
